src/richie/apps/courses/models/organization.py

```diff
@@ -167,6 +169,50 @@ def get_persons(self, language=None):
         """
         return self.get_reverse_related_page_extensions("person", language=language)
 
+    @staticmethod
+    def get_organizations_codes(page, language: str):
+        """
+        Return organization attached to the page and
+        organizations linked to the current page via an organization plugin in any of the
+        placeholders on the page.
+        This method is cached, so it's possible to return inconsistent information after a
+        page has been changed.
+        """
+        cache_key = (
+            f"richie_apps_courses_models_organization_"
+            f"get_organizations_codes_{page.id:d}_{language:s}"
+        )
+        codes = cache.get(cache_key)
+
+        if not codes:
+            codes = (
+                Organization.objects.filter(
+                    Q(
+                        extended_object__organization_plugins__cmsplugin_ptr__language=language,
+                        extended_object__organization_plugins__cmsplugin_ptr__placeholder__page=page,  # noqa pylint: disable=line-too-long
+                    )
+                    | Q(
+                        extended_object__exact=page,
+                    ),
+                    extended_object__title_set__published=True,
+                    code__isnull=False,
+                )
+                .distinct()
+                .order_by("code")
+                .values_list("code", flat=True)
+            )
+            # Cache for one 6 minutes the organization codes
+            cache.set(
+                cache_key,
+                codes,
+                getattr(
+                    settings,
+                    "RICHIE_COURSES_ORGANIZATION_CODES_CACHE_TIMEOUT_SECONDS",
+                    600,
+                ),
+            )
+        return codes
+
```

Comment:
Not sure that the result of this method should be cached. As you explain, it could lead to stale data. From my point of view, it's ok to just test that once a page is cached, db queries related to this method are not executed again. In this way, if a page is edited, cache will be flush and data will be always fresh.

You can check those lines to understand what I mean
https://github.com/openfun/richie-site-factory/blob/cffb0871d940297c5784f9350f7067c1f2530a5e/sites/funmooc/src/backend/base/tests/test_context_processors.py#L117-L135

Reply:
What do you mean by "once a page is cached" ? On the client browser? Or on Django cache?

For multiple clients the same page would be requested multiple times, at least one per client. Then the app would execute the expensive SQL for each request ( or the Django page cache resolves this? ).

The organizations of a page probably should be changed only a few times. I think some stale data on web analytics organizations during a couple of minutes is not a problem.

Reply:
I mean by the Django cache, so once a page has been cached, clients retrieves this cached verison.

Reply:
Ok, now I understand. I removed the cache on this method.